### ISSUE TITLE
Cyberduck not listen files

### ISSUE THREAD

**Issue by francoricci** [leo-project/leofs#1116]
Until version 6.2.2 cyberduck listen files inside leofs buckets and directories.
Latest version of cyberduck can list only directories but now files inside is accessible.

**Comment by mocchira** [comment]
@francoricci Thanks for reporting this issue. I've confirmed on the latest (6.7.1.28683). Since other s3 clients works fine, there might be something wrong specific to the cyberduck latest. I'll dive into the cyberduck code from now on. Until I find the fix, would you like to keep using 6.2.2 cyberduck or try dragondisk (http://www.s3-client.com/) out as a GUI browser?

**Comment by mocchira** [comment]
Found the differences between the one succeeding and the another failing.

- The succeeding one has non-empty NextMarker field
- The failing one receives HTTP Body in HTTP/1.0 way (Connection: close) OHOH the succeeding one receives in HTTP/1.1 way (ContentLength is set to HTTP Header)

### Example Request/Response
- Succeeded
```bash
*   Trying 192.168.100.10...
* Connected to 192.168.100.10 (192.168.100.10) port 18443 (#0)
* found 148 certificates in /etc/ssl/certs/ca-certificates.crt
* found 592 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / ECDHE_RSA_AES_128_GCM_SHA256
*        server certificate verification SKIPPED
*        server certificate status verification SKIPPED
*        common name: localhost (does not match '192.168.100.10')
*        server certificate expiration date OK
*        server certificate activation date OK
*        certificate public key: RSA
*        certificate version: #3
*        subject: CN=localhost
*        start date: Thu, 18 Mar 2010 19:39:18 GMT
*        expire date: Sun, 15 Mar 2020 19:39:18 GMT
*        issuer: CN=localhost
*        compression: NULL
* ALPN, server did not agree to a protocol
> GET /test/?max-keys=1000&versions&prefix=hoge%2F&delimiter=%2F HTTP/1.1
> Host: 192.168.100.10
> Accept: */*
> Date: Tue, 04 Sep 2018 07:03:38 GMT
> x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
> x-amz-date: 20180904T070338Z
> Authorization: AWS4-HMAC-SHA256 Credential=05236/20180904/us-east-1/s3/aws4_request,SignedHeaders=date;host;x-amz-content-sha256;x-amz-date,Signature=3c26f62f34fb713d561a7d2bfa3f3a23e144ccf1beb8b98a52e651af62e
e5824
> Connection: Keep-Alive
> User-Agent: Cyberduck/6.7.1.28683 (Windows 10/10.0) (x86)
>
< HTTP/1.1 200 OK
< Connection: keep-alive
< Date: Tue, 04 Sep 2018 07:29:20 GMT
< Content-Length: 828
< Server: LeoFS
< Content-Type: application/xml
<
* Connection #0 to host 192.168.100.10 left intact
<?xml version="1.0" encoding="UTF-8"?><ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Name>test</Name><Prefix>hoge/</Prefix><Marker></Marker><MaxKeys>1000</MaxKeys><Delimiter>/</Delimiter><Con
tents><Key>hoge/Achievement_2018_4.pdf</Key><LastModified>2018-09-04T01:28:39.000Z</LastModified><ETag>f5cd92eb7ab6e3b453268bab05e21928</ETag><Size>53885</Size><StorageClass>STANDARD</StorageClass><Owner><ID>leo
fs</ID><DisplayName>leofs</DisplayName></Owner></Contents><Contents><Key>hoge/rebar</Key><LastModified>2018-09-04T01:29:55.000Z</LastModified><ETag>3178fff900863cb8847d211a2dce053e</ETag><Size>181045</Size><Stor
ageClass>STANDARD</StorageClass><Owner><ID>leofs</ID><DisplayName>leofs</DisplayName></Owner></Contents><IsTruncated>false</IsTruncated><NextMarker>test/hoge/rebar</NextMarker></ListBucketResult>
```
- Failed
```bash
*   Trying 192.168.100.10...
* Connected to 192.168.100.10 (192.168.100.10) port 18443 (#0)
* found 148 certificates in /etc/ssl/certs/ca-certificates.crt
* found 592 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / ECDHE_RSA_AES_128_GCM_SHA256
*        server certificate verification SKIPPED
*        server certificate status verification SKIPPED
*        common name: localhost (does not match '192.168.100.10')
*        server certificate expiration date OK
*        server certificate activation date OK
*        certificate public key: RSA
*        certificate version: #3
*        subject: CN=localhost
*        start date: Thu, 18 Mar 2010 19:39:18 GMT
*        expire date: Sun, 15 Mar 2020 19:39:18 GMT
*        issuer: CN=localhost
*        compression: NULL
* ALPN, server did not agree to a protocol
> GET /test/?prefix=hoge%2F&uploads HTTP/1.1
> Host: 192.168.100.10
> Accept: */*
> Date: Tue, 04 Sep 2018 07:32:15 GMT
> x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
> x-amz-date: 20180904T073215Z
> Authorization: AWS4-HMAC-SHA256 Credential=05236/20180904/us-east-1/s3/aws4_request,SignedHeaders=date;host;x-amz-content-sha256;x-amz-date,Signature=9c23481bb61e6b58d66eef758ac64e633b9d118953346358275dbd66522
5d873
> Connection: Keep-Alive
> User-Agent: Cyberduck/6.7.1.28683 (Windows 10/10.0) (x86)
>
< HTTP/1.1 200 OK
< Transfer-Encoding: identity
< Date: Tue, 04 Sep 2018 07:34:44 GMT
< Connection: close
< Server: LeoFS
< Content-Type: application/xml
<
* Closing connection 0
<?xml version="1.0" encoding="UTF-8"?><ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Name>test</Name><Prefix>hoge/</Prefix><Marker></Marker><MaxKeys>1000</MaxKeys><Delimiter>/</Delimiter><Con
tents><Key>hoge/Achievement_2018_4.pdf</Key><LastModified>2018-09-04T01:28:39.000Z</LastModified><ETag>f5cd92eb7ab6e3b453268bab05e21928</ETag><Size>53885</Size><StorageClass>STANDARD</StorageClass><Owner><ID>leo
fs</ID><DisplayName>leofs</DisplayName></Owner></Contents><Contents><Key>hoge/rebar</Key><LastModified>2018-09-04T01:29:55.000Z</LastModified><ETag>3178fff900863cb8847d211a2dce053e</ETag><Size>181045</Size><Stor
ageClass>STANDARD</StorageClass><Owner><ID>leofs</ID><DisplayName>leofs</DisplayName></Owner></Contents><IsTruncated>false</IsTruncated><NextMarker></NextMarker></ListBucketResult>
```

**Comment by francoricci** [comment]
I'll use the old version f CyberDuck until solution.
Thanks

**Comment by mocchira** [comment]
The culprit function which return the response body in HTTP/1.0 way
- https://github.com/leo-project/leo_gateway/blame/develop/src/leo_gateway_s3_api.erl#L1603-L1638

- Related commits
  - https://github.com/leo-project/leo_gateway/commit/01a653b35429541ce3bb64c777bcf4521bef8430
  - https://github.com/leo-project/leo_gateway/commit/3cc5866fdf78b51f99abea2849a9b60f6a241039

**Comment by mocchira** [comment]
Memo: diff between 6.2.2 and 6.7.2 (the latest): https://trac.cyberduck.io/changeset?old_path=%2Ftags%2Frelease-6-2-2%2Fs3%2Fsrc%2Fmain%2Fjava%2Fch%2Fcyberduck%2Fcore%2Fs3%2FS3ObjectListService.java&old=45028&new_path=%2Ftags%2Frelease-6-7-2%2Fs3%2Fsrc%2Fmain%2Fjava%2Fch%2Fcyberduck%2Fcore%2Fs3%2FS3ObjectListService.java&new=45028

**Comment by mocchira** [comment]
It turns out that several comments I previously posted at here were wrong so I deleted those to keep comments clean.

**Comment by mocchira** [comment]
The reason why the latest cyberduck can't display any files but display folders stored in LeoFS is that they've changed the API retrieving the object list from https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGET.html to https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGETVersion.html.

Since LeoFS has no yet supported **GET Bucket Object versions** API, any files have not been displayed on the latest cyberduck. At a first glance, the simple implementation which just returns the most recent objects can be implemented in quite less cost so that I'm going to implement it in that way and confirm if it works as expected.

**Comment by mocchira** [comment]
@francoricci I've just sent PR https://github.com/leo-project/leofs/pull/1118 to fix the problem. The next stable release 1.4.3 will include the fix so once 1.4.3 comes out and you finish to upgrade LeoFS, you will be able to use the latest cyberduck. Please wait for a while.

**Comment by francoricci** [comment]
Great!
Thanks a lot, I'll wait for the next leofs release.

Franco
